src/compiler/parser.ts

```diff
@@ -2717,6 +2745,31 @@ module ts {
             return expr;
         }
 
+        function parseBindingElementInitializer(inParameter: boolean): Expression {
+            // BindingElement[Yield,GeneratorParameter,Await,AsyncParameter] : 
+            //      [+GeneratorParameter] BindingPattern[?Yield,?Await,GeneratorParameter] Initializer[In]opt
+            //      [+AsyncParameter] BindingPattern[?Yield,?Await,AsyncParameter] Initializer[In]opt
+            //      [~GeneratorParameter,~AsyncParameter] BindingPattern[?Yield,?Await] Initializer[In,?Yield,?Await]opt
+            // SingleNameBinding[Yield,GeneratorParameter,Await,AsyncParameter] : 
+            //      [+GeneratorParameter] BindingIdentifier[Yield] Initializer[In]opt
+            //      [+AsyncParameter] BindingIdentifier[Await] Initializer[In]opt
+            //      [~GeneratorParameter,~AsyncParameter] BindingIdentifier[?Yield,?Await] Initializer[In,?Yield,?Await]opt
+            var explicitFlags = ParserContextFlags.None;
+            var ambientFlags = ParserContextFlags.None;
+            if (!inGeneratorParameterContext() && !inAsyncParameterContext()) {
+                ambientFlags = ParserContextFlags.Yield | ParserContextFlags.Await;
+            }
+            return setContextParametersAnd(inParameter ? parseParameterInitializer : parseNonParameterInitializer, explicitFlags, ambientFlags);
+        }
+
+        function parseParameterInitializer() {
+            return parseInitializer(/*inParameter*/ true);

```

Comment:
See the comment in `parseInitializer`, the parameter is used for error recovery.

From: Jason Freeman [mailto:notifications@github.com]
Sent: Thursday, February 12, 2015 3:09 PM
To: Microsoft/TypeScript
Cc: Ron Buckton
Subject: Re: [TypeScript] Async Functions (#1781)


In src/compiler/parser.ts<https://github.com/Microsoft/TypeScript/pull/1781#discussion_r24630433>:

> +            //      [+AsyncParameter] BindingPattern[?Yield,?Await,AsyncParameter] Initializer[In]opt

> +            //      [~GeneratorParameter,~AsyncParameter] BindingPattern[?Yield,?Await] Initializer[In,?Yield,?Await]opt

> +            // SingleNameBinding[Yield,GeneratorParameter,Await,AsyncParameter] :

> +            //      [+GeneratorParameter] BindingIdentifier[Yield] Initializer[In]opt

> +            //      [+AsyncParameter] BindingIdentifier[Await] Initializer[In]opt

> +            //      [~GeneratorParameter,~AsyncParameter] BindingIdentifier[?Yield,?Await] Initializer[In,?Yield,?Await]opt

> +            var explicitFlags = ParserContextFlags.None;

> +            var ambientFlags = ParserContextFlags.None;

> +            if (!inGeneratorParameterContext() && !inAsyncParameterContext()) {

> +                ambientFlags = ParserContextFlags.Yield | ParserContextFlags.Await;

> +            }

> +            return setContextParametersAnd(inParameter ? parseParameterInitializer : parseNonParameterInitializer, explicitFlags, ambientFlags);

> +        }

> +

> +        function parseParameterInitializer() {

> +            return parseInitializer(/*inParameter*/ true);

Why do we need to know if we're in a parameter?

—
Reply to this email directly or view it on GitHub<https://github.com/Microsoft/TypeScript/pull/1781/files#r24630433>.